### PR TITLE
Implemention of SpaceSaving algorithm for TopN

### PR DESCRIPTION
**Pull request by WireBaron**
This adds a TopN implementation to Postgres using the SpaceSaving algorithm.

Currently this is only implemented for integer data, and should be generalized to any type before it's ready for being stabilized.

### CODE REVIEW

File: extension/src/topn.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn topn_iter (
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
+) -> impl std::iter::Iterator<Item = (name!(value,i64),name!(min_freq,f64),name!(max_freq,f64))> + '_ {
+    assert!(n <= agg.num_values as _);
+    let mut count = 0;
+    let total = agg.total_inputs as f64;
+    Box::new(multizip((agg.values.iter(), agg.counts.iter(), agg.overcounts.iter())).filter(move |_| {count = count + 1; count <= n}).map(move |x| {let (val, count, over) = x; (*val, (count-over) as f64 / total, *count as f64 / total)}))
```
Comment (JLockerman) [inline extension/src/topn.rs:214]:
```suggestion
    multizip((agg.values.iter(), agg.counts.iter(), agg.overcounts.iter())).take(n).map(move |x| {let (val, count, over) = x; (*val, (count-over) as f64 / total, *count as f64 / total)})
```

File: extension/src/topn.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+pub fn num_vals(
+    agg: toolkit_experimental::TopN<'static>,
+) -> int {
+    agg.total_inputs as _
+}
+
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn topn_iter (
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
```
Comment (JLockerman) [inline extension/src/topn.rs:209]:
```suggestion
    agg: toolkit_experimental::TopN<'input>,
```

This should _never_ be `'static` or the same as the output; that claims that the input argument will live forever when it only lives for this one call.

**Review by JLockerman — CHANGES_REQUESTED** [review]

File: extension/src/topn.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    assert!(n <= agg.num_values as _);
+    let mut count = 0;
+    let total = agg.total_inputs as f64;
+    Box::new(multizip((agg.values.iter(), agg.counts.iter(), agg.overcounts.iter())).filter(move |_| {count = count + 1; count <= n}).map(move |x| {let (val, count, over) = x; (*val, (count-over) as f64 / total, *count as f64 / total)}))
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn guaranteed_topn(
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
```
Comment (JLockerman) [inline extension/src/topn.rs:220]:
```suggestion
pub fn guaranteed_topn<'input>(
    n: i32,
    agg: toolkit_experimental::TopN<'input>,
```

File: extension/src/topn.rs [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            return false;
+        }
+    }
+
+    true
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn max_ordered_n(
+    agg: toolkit_experimental::TopN<'static>,
```
Comment (JLockerman) [inline extension/src/topn.rs:238]:
```suggestion
pub fn max_ordered_n<'input>(
    agg: toolkit_experimental::TopN<'input>,
```

File: extension/src/topn.rs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+//---- Available PG operations on the topn structure
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn num_vals(
+    agg: toolkit_experimental::TopN<'static>,
```
Comment (JLockerman) [inline extension/src/topn.rs:201]:
```suggestion
pub fn num_vals<'input>(
    agg: toolkit_experimental::TopN<'input>,
```

File: extension/src/topn.rs [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn topn_compound_trans(
+    state: Option<Internal<InternalTopN>>,
+    value: Option<toolkit_experimental::TopN<'static>>,
```
Comment (JLockerman) [inline extension/src/topn.rs:169]:
```suggestion
pub fn topn_compound_trans<'b>(
    state: Option<Internal<InternalTopN>>,
    value: Option<toolkit_experimental::TopN<'b>>,
```

File: extension/Cargo.toml [hunk 7]
```diff
@@ -37,6 +38,7 @@ paste = "1.0"
 rand = { version = "0.8.3", features = ["getrandom", "small_rng"] }
 rand_distr = "0.4.0"
 rand_chacha = "0.3.0"
+itertools = "0.7.8"
```
Comment (JLockerman) [inline extension/Cargo.toml:41]:
```suggestion
itertools = "0.10.1"
```

File: extension/src/topn.rs [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    agg: toolkit_experimental::TopN<'static>,
+) -> int {
+    for i in 1..agg.num_values as usize {
+        if agg.counts[i] > agg.counts[i-1] - agg.overcounts[i-1] {
+            return (i - 1) as _;
+        }
+    }
+
+    (agg.num_values - 1) as _
+}
```
Comment (JLockerman) [inline extension/src/topn.rs:247]: Note for the future: we probably want a `top_guaranteed()` which outputs elements until this loop condition is false.

File: extension/src/topn.rs (hunk 2 above)
Comment (WireBaron) [reply extension/src/topn.rs:220]: This actually couldn't be made non-static without breaking the compilation.  After discussing this we've come up with an explanation of why this is okay, which has been added as a comment here.

File: extension/src/topn.rs (hunk 8 above)
Comment (WireBaron) [reply extension/src/topn.rs]: Agreed.

**Pull request merged**